src/client/opamAuxCommands.ml

```diff
@@ -303,16 +303,31 @@ let autopin_aux st ?quiet ?(for_view=false) ?recurse ?subpath ?locked
               | _ -> false)
            | None -> false)
           &&
-          (* For `opam show`, we need to check does the opam file changed to
-             perform a simulated pin if so *)
-          (not for_view ||
-           match
+          (match
              OpamSwitchState.opam_opt st pinned_pkg,
              OpamFile.OPAM.read_opt nf.pin.pin_file
            with
            | Some opam0, Some opam ->
              let opam = OpamFile.OPAM.with_locked_opt nf.pin.pin_locked opam in
-             OpamFile.OPAM.equal opam0 opam
+             let opam = OpamFile.OPAM.with_name nf.pin_name opam in
+             let opam =
+               match OpamFile.OPAM.version_opt opam with
+               | None ->
+                 OpamFile.OPAM.with_version
+                   (OpamPinCommand.default_version st nf.pin_name) opam
+               | Some _ -> opam
+             in
+             let opam =
+               OpamFile.OPAM.with_url
+                 (OpamFile.URL.create ?subpath:nf.pin.pin_subpath nf.pin.pin_url)
+                 opam
+             in
+             let opam =
+               (* This is required to avoid the case where locked opam files were
+                  stored with the added `available: opam-version >= "2.1"` *)
+               OpamFile.OPAM.with_available (OpamFile.OPAM.available opam0) opam
```

Comment:
It worth adding a comment where these fields are added for the internal stored opam, in order to update this piece of code when it is changed.
I don't know if we can have a test to follow that, keep the consistency between these fields added, and the ones checked here.